src/neo4j_genai/core/pipeline_graph.py

```diff
@@ -0,0 +1,135 @@
+
+    def add_node(self, node: PipelineNode) -> None:
+        if node in self:
+            raise ValueError(f"Node {node.name} already exists")
+        self._nodes[node.name] = node
+
+    def set_node(self, node: PipelineNode) -> None:
+        """Replace an existing node with a new one based on node name."""
+        if node not in self:
+            raise ValueError(f"Node {node.name} does not exist")
+        # propagate the graph info to the new node:
+        old_node = self._nodes[node.name]
+        node.parents = old_node.parents
+        node.children = old_node.children
+        self._nodes[node.name] = node
+
+    def _validate_edge(self, start: str, end: str) -> None:
+        if start not in self:
+            raise ValueError(f"Node {start} does not exist")
+        if end not in self:
+            raise ValueError(f"Node {end} does not exist")
+        for edge in self._edges:
+            if edge.start == start and edge.end == end:
+                raise ValueError(f"{start} and {end} are already connected")
+
+    def connect(
+        self, start: PipelineNode, end: PipelineNode, data: dict[str, Any]
+    ) -> None:
+        self._validate_edge(start.name, end.name)
+        self._edges.append(PipelineEdge(start.name, end.name, data))
+        self._nodes[end.name].parents.append(start.name)
+        self._nodes[start.name].children.append(end.name)
+
+    def get_node_by_name(
+        self, name: str, raise_exception: bool = False
+    ) -> PipelineNode:
+        node = self._nodes.get(name)
+        if node is None and raise_exception:
+            raise KeyError(f"Component {name} not in graph")
+        return node  # type: ignore
+
+    def roots(self) -> list[PipelineNode]:
+        root = []
+        for node in self._nodes.values():
+            if node.is_root():
+                root.append(node)
+        return root
+
+    def next_edges(self, node: str) -> list[PipelineEdge]:
+        res = []
+        for edge in self._edges:
+            if edge.start == node:
+                res.append(edge)
+        return res
+
+    def previous_edges(self, node: str) -> list[PipelineEdge]:
+        res = []
+        for edge in self._edges:
+            if edge.end == node:
+                res.append(edge)
+        return res
+
+    def __contains__(self, node: PipelineNode | str) -> bool:
+        if isinstance(node, PipelineNode):
+            return node.name in self._nodes
+        return node in self._nodes
+
+    def dfs(self, visited: set[PipelineNode], node: PipelineNode) -> bool:
+        if node in visited:
+            return True
+        else:
+            for edge in self.next_edges(node.name):
+                neighbour = self.get_node_by_name(edge.end)
+                if self.dfs(visited | {node}, neighbour):
+                    return True
+            return False
+
+    def is_cyclic(self) -> bool:
+        for node in self._nodes.values():
+            if self.dfs(set(), node):
```

Comment:
IIUC the way we check for cycles is if the dfs visits a node with a seen name, it returns False. Perhaps it will be good to document this function

Reply:
Done in https://github.com/neo4j/neo4j-genai-python/pull/81/commits/e5a7d292c3c853587bbd7d92d7f84fe43ea87c59 (also added tests)